src/server/accept.rs

```diff
@@ -64,7 +52,11 @@ where
     // of connections that the server can handle concurrently. If the maximum
     // number of connections is reached, we'll wait until a permit is available
     // before accepting a new connection.
-    let semaphore = Arc::new(Semaphore::new(max_connections));
+    let semaphore = Arc::new(Semaphore::new(context.max_connections()));
+
+    // Create a JoinSet to track inflight connections. We'll use this to wait for
+    // all connections to close before the server exits.
+    let mut connections = VecDeque::<TaskHandle>::with_capacity(4096);
```

Comment:
I'm a fan of the big allocation here with `VecDeque`. I think we're smart enough to guess the allocation size required to support the server even if it's at capacity.

This is one of the only reasons why I would prefer using a VecDeque rather than `JoinSet`. The implementation of `JoinSet` is great but I'd prefer that no dynamic allocations occur in the accept loop.

I'm considering submitting a PR to tokio that adds a `with_capacity` method to `JoinSet` for this reason.